src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs

```diff
@@ -2893,9 +3574,322 @@ static S2 Test3()
+
+                        S1 local;
+                        local = new S1() { [stackSpan] =  0, [heapSpan] = 1 };
+                    }
+                }
+
+                public ref struct S2
+                {
+                    ref int this[scoped Span<int> span] => throw null!;
+
+                    static void Test()
+                    {
+                        Span<int> stackSpan = stackalloc int[] { 13 };
+                        Span<int> heapSpan = default;
+
+                        S2 local;
+                        local = new S2() { [stackSpan] =  0, [heapSpan] = 1 };
+                    }
+                }
+                """, TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp11));
+
+            var comp = CreateCompilationWithSpan(tree, TestOptions.UnsafeDebugDll);
+            comp.VerifyEmitDiagnostics();
+        }
+
+        [Fact]
+        public void ObjectInitializer_Indexer_RefLikeType()
+        {
+            var tree = SyntaxFactory.ParseSyntaxTree("""
+                using System;
+
+                public ref struct S1
+                {
+                    Span<int> this[int index] { get => default; set { } }
+
+                    static void Test()
+                    {
+                        Span<int> stackSpan = stackalloc int[] { 13 };
+                        Span<int> heapSpan = default;
+
+                        S1 local;
+                        local = new S1() { [0] =  stackSpan }; // 1
+                        local = new S1() { [0] =  heapSpan };
+                    }
+                }
+
+                public ref struct S2
+                {
+                    Span<int> this[int index] { get => default; readonly set { } }
+
+                    static void Test()
+                    {
+                        Span<int> stackSpan = stackalloc int[] { 13 };
+                        Span<int> heapSpan = default;
+
+                        S2 local;
+                        local = new S2() { [0] =  stackSpan };
+                        local = new S2() { [0] =  heapSpan };
+                    }
+                }
+                """, TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp11));
+
+            var comp = CreateCompilationWithSpan(tree, TestOptions.UnsafeDebugDll);
+            comp.VerifyEmitDiagnostics(
+                // (13,26): error CS8352: Cannot use variable '[0] =  stackSpan' in this context because it may expose referenced variables outside of their declaration scope
+                //         local = new S1() { [0] =  stackSpan }; // 1
+                Diagnostic(ErrorCode.ERR_EscapeVariable, "{ [0] =  stackSpan }").WithArguments("[0] =  stackSpan").WithLocation(13, 26));
+        }
+
+        [Fact]
+        public void ObjectInitializer_Constructor_Escape()
+        {
+            var tree = SyntaxFactory.ParseSyntaxTree("""
+                using System;
+
+                public ref struct S1
```

Comment:
>public ref struct S1 [](http://example.com/codeflow?start=16&length=20)

Are we testing the same example but with `struct S1` rather than `ref struct S1`, where the object initializers should not generate any errors?

Reply:
Good catch.